File: python/boj/1920.py
import sys
from typing import List

input = sys.stdin.readline


def in_n(target: int, n_list: List[int]) -> bool:
    def binary_search(left: int, right: int) -> bool:
        if left <= right:
            mid = left + (right - left) // 2
            if n_list[mid] < target:
                return binary_search(mid + 1, right)
            elif n_list[mid] > target:
                return binary_search(left, mid - 1)
            else:
                return True
        return False

    return binary_search(0, len(n_list) - 1)


def solution():
    n = int(input())
    n_list = list(map(int, input().split()))
    n_list.sort()

    m = int(input())
    m_list = list(map(int, input().split()))

    for num in m_list:
        if in_n(num, n_list):
            print(1)
        else:
            print(0)


if __name__ == "__main__":
    solution()

File: cpp/bj/1920.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;


int binarySearch(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size() - 1;

    while (left <= right) {
        int mid = right + (left - right) / 2;
        if (nums[mid] < target) {
            left = mid + 1;
        } else if (nums[mid] > target) {
            right = mid - 1;
        } else {
            return mid;
        }
    }

    return -1;
}

bool hasNum(vector<int>& nums, int target) {
    int idx = binarySearch(nums, target);
    if (idx != -1) return true;
    return false;
}

void solution() {
    cin.tie(NULL);
    cout.tie(NULL);
    std::ios::sync_with_stdio(false);
    int n, m;
    cin >> n;

    vector<int> nums;

    for (int i = 0; i < n; i++) {
        int num;
        cin >> num;
        nums.push_back(num);
    }
    sort(nums.begin(), nums.end());

    int input;
    cin >> m;
    for (int i = 0; i < m; i++) {
        cin >> input;
        if (hasNum(nums, input) == true) {
            cout << 1 << '\n';
        } else {
            cout << 0 << '\n';
        }
    }
}


int main() {
    solution();
    return 0;
}